### PR TITLE
Rework attribute generator and use now registry entries

### PR DESCRIPTION
**Pull request by TheMeinerLP**
## Proposed changes

This PR serves the purpose of better generating the attributes. It uses the standard registry system from Minestom to avoid errors and increase consistency in the code. I have discarded the original design and added two new classes. One for the entry from the file for the registry and one for the interface implementation to add the data as getter etc. The attribute class became an interface to follow the structure. The enum generation I created before was replaced with an interface and the enum generator was deleted as a class.

In the end you can now call the class via static attributes as before but update it faster in the future.

## Types of changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [X] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before
merging your code._

- [X] I have read the CONTRIBUTING.md
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you
did and what alternatives you considered, etc...

### CODE REVIEW

File: minestom-patches/0015-Add-attribute-generator.patch [hunk 1]
```diff
@@ -50,56 +49,61 @@ index 618040b6ff868f2327c641ed3e5c205c8a00344f..2bd0ec47837841bf2654fdacd8524ec8
 -import com.squareup.javapoet.ParameterizedTypeName;
 -import com.squareup.javapoet.MethodSpec;
 -import com.squareup.javapoet.ParameterSpec;
-+import com.squareup.javapoet.*;
```
Comment (theEvilReaper) [inline minestom-patches/0015-Add-attribute-generator.patch:53]: Don't use wildcard imports there

File: minestom-patches/0015-Add-attribute-generator.patch [hunk 2]
```diff
@@ -607,35 +556,118 @@ index ed97a493c48bc213077bababac2723562b53dc6f..0b1a3ceb6e6f8fab6c7b66e34b825480
...
+@@ -0,0 +1,74 @@
++package net.minestom.server.attribute;
++
++import net.minestom.server.registry.ProtocolObject;
++import net.minestom.server.registry.Registry;
++import net.minestom.server.utils.NamespaceID;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.Collection;
++
```
Comment (theEvilReaper) [inline minestom-patches/0015-Add-attribute-generator.patch:576]: We should annotate this with the `ApiStatus.NonExtendable` annotation to indicate that the class can't be extended

File: minestom-patches/0015-Add-attribute-generator.patch [hunk 3]
```diff
@@ -607,35 +556,118 @@ index ed97a493c48bc213077bababac2723562b53dc6f..0b1a3ceb6e6f8fab6c7b66e34b825480
...
++    @NotNull
++    private final NamespaceID name;
++    private final float defaultValue;
++    private final float maxValue;
++    private final float minValue;
++    private final boolean clientSync;
++    @NotNull
++    private final String translationKey;
++
++    public AttributeImpl(Registry.AttributeEntry entry) {
```
Comment (theEvilReaper) [inline minestom-patches/0015-Add-attribute-generator.patch:603]: Add `NotNull` annotation

File: minestom-patches/0015-Add-attribute-generator.patch [hunk 4]
```diff
@@ -607,35 +556,118 @@ index ed97a493c48bc213077bababac2723562b53dc6f..0b1a3ceb6e6f8fab6c7b66e34b825480
...
++    private final Map<NamespaceID, Attribute> attributes = new ConcurrentHashMap<>();
 +
 +    /**
 +     * Register this attribute.
 +     *
-+     * @see #fromKey(String)
++     * @see #fromKey(NamespaceID)
 +     * @see #values()
 +     */
 +    public synchronized void register(Attribute attribute) {
```
Comment (theEvilReaper) [inline minestom-patches/0015-Add-attribute-generator.patch:669]: Add `NotNull` annotation

File: minestom-patches/0015-Add-attribute-generator.patch [hunk 5]
```diff
@@ -607,35 +556,118 @@ index ed97a493c48bc213077bababac2723562b53dc6f..0b1a3ceb6e6f8fab6c7b66e34b825480
...
 +import java.util.Collection;
 +
 +import java.util.Map;
 +import java.util.concurrent.ConcurrentHashMap;
 +
++import net.minestom.server.utils.NamespaceID;
 +import org.jetbrains.annotations.NotNull;
 +import org.jetbrains.annotations.Nullable;
 +
 +public final class AttributeManager {
```
Comment (theEvilReaper) [inline minestom-patches/0015-Add-attribute-generator.patch:659]: Also here annotate the class with `ApiStatus.NonExtendable`

File: minestom-patches/0015-Add-attribute-generator.patch [hunk 6]
```diff
@@ -801,10 +833,107 @@ index 98fefc2ee8530bd753b6271b4d82b040711d49d7..fea7090fe4f672cdc089d17edc8d942e
...
++    public static final class AttributeEntry implements Entry {
++        private final Properties custom;
++        private final NamespaceID namespace;
++        private final float defaultValue;
++        private final String translationKey;
++        private final boolean clientSync;
++        private final float maxValue;
++        private final float minValue;
++
++        public AttributeEntry(String namespace, Properties main, Properties custom) {
```
Comment (theEvilReaper) [inline minestom-patches/0015-Add-attribute-generator.patch:889]: Annotate the parameters with `NotNull` or `Nullable` depends on the case

File: minestom-patches/0015-Add-attribute-generator.patch [hunk 7]
```diff
@@ -570,13 +502,26 @@ index ed97a493c48bc213077bababac2723562b53dc6f..0b1a3ceb6e6f8fab6c7b66e34b825480
...
-     }
 -
+-    public Attribute {
+-        if (defaultValue > maxValue) {
+-            throw new IllegalArgumentException("Default value cannot be greater than the maximum allowed");
+-        }
++public sealed interface Attribute extends ProtocolObject, Attributes permits AttributeImpl {
++
++    @Override
++    @NotNull
```
Comment (theEvilReaper) [inline minestom-patches/0015-Add-attribute-generator.patch:516]: We can move the `NotNull` annotation before the NameSpaceID definition. So the line looks like this `NotNull NameSpaceID namespace();` (Dont forget to add the annotation char. I don't want to tag someone here)

File: minestom-patches/0015-Add-attribute-generator.patch [hunk 8]
```diff
@@ -570,13 +502,26 @@ index ed97a493c48bc213077bababac2723562b53dc6f..0b1a3ceb6e6f8fab6c7b66e34b825480
...
++    NamespaceID namespace();
++    float defaultValue();
++    String translationKey();
++    boolean isClientSync();
++    float maxValue();
++    float minValue();
```
Comment (theEvilReaper) [inline minestom-patches/0015-Add-attribute-generator.patch]: Add between each getter an empty line to increate readability. Also some annotation would be nice

**Review by theEvilReaper — CHANGES_REQUESTED** [review]

File: minestom-patches/0015-Add-attribute-generator.patch (hunk 8 above)
Comment (TheMeinerLP) [reply minestom-patches/0015-Add-attribute-generator.patch]: Thats not longer valid because outdated view from you.
image

**Pull request merged**